cpp/src/arrow/compute/kernels/scalar_cast_temporal.cc

```diff
@@ -510,9 +510,12 @@ void AddCrossUnitCastNoPreallocate(CastFunction* func) {
 
 std::shared_ptr<CastFunction> GetDate32Cast() {
   auto func = std::make_shared<CastFunction>("cast_date32", Type::DATE32);
-  auto out_ty = date32();
+  const auto& out_ty = date32();
   AddCommonCasts(Type::DATE32, out_ty, func.get());
 
+  // date32 -> date32
+  AddCrossUnitCast<Date32Type>(func.get());
```

Comment:
Can't we just call `AddZeroCopyCast` here? Only a single unit is supported for date32.